File: frontend/lib/recommender.ts
import {
  NUTRIENT_CONFLICTS,
  NUTRIENT_DESCRIPTION,
  NUTRIENT_SYNERGIES,
  NUTRIENT_TIMING,
  PRODUCTS,
  SYMPTOM_TO_NUTRIENT,
  type Product,
  type Timing,
} from "./data";

export type RecommendedNutrient = {
  name: string;
  description: string;
  timing: Timing | null;
};
export type RecommendedProduct = Product & { nutrient: string };
export type PairInsight = { a: string; b: string; reason: string };

export type RecommendResult = {
  nutrients: RecommendedNutrient[];
  products: RecommendedProduct[];
  synergies: PairInsight[];
  conflicts: PairInsight[];
};

export function collectNutrients(symptoms: string[]): string[] {
  const seen = new Set<string>();
  const ordered: string[] = [];
  for (const symptom of symptoms) {
    for (const nutrient of SYMPTOM_TO_NUTRIENT[symptom] ?? []) {
      if (seen.has(nutrient)) continue;
      seen.add(nutrient);
      ordered.push(nutrient);
    }
  }
  return ordered;
}

export function collectProducts(nutrients: string[]): RecommendedProduct[] {
  const seen = new Set<string>();
  const result: RecommendedProduct[] = [];
  for (const nutrient of nutrients) {
    for (const product of PRODUCTS[nutrient] ?? []) {
      if (seen.has(product.name)) continue;
      seen.add(product.name);
      result.push({ ...product, nutrient });
    }
  }
  return result;
}

function findPairsAmong(
  nutrients: string[],
  entries: Array<{ pair: [string, string]; reason: string }>,
): PairInsight[] {
  const set = new Set(nutrients);
  return entries
    .filter(({ pair }) => set.has(pair[0]) && set.has(pair[1]))
    .map(({ pair, reason }) => ({ a: pair[0], b: pair[1], reason }));
}

export function recommend(symptoms: string[]): RecommendResult {
  const nutrients = collectNutrients(symptoms);
  return {
    nutrients: nutrients.map((name) => ({
      name,
      description: NUTRIENT_DESCRIPTION[name] ?? "",
      timing: NUTRIENT_TIMING[name] ?? null,
    })),
    products: collectProducts(nutrients),
    synergies: findPairsAmong(nutrients, NUTRIENT_SYNERGIES),
    conflicts: findPairsAmong(nutrients, NUTRIENT_CONFLICTS),
  };
}

export function listSymptoms(): string[] {
  return Object.keys(SYMPTOM_TO_NUTRIENT);
}

File: backend/app/recommender.py
"""증상 입력을 받아 영양소와 제품을 추천하는 핵심 로직."""

from __future__ import annotations

from collections.abc import Iterable

from .data import (
    NUTRIENT_DESCRIPTION,
    PRODUCTS,
    SYMPTOM_TO_NUTRIENT,
    Product,
)


def collect_nutrients(symptoms: Iterable[str]) -> list[str]:
    """증상 리스트에서 중복 없이 영양소를 모은다.

    입력 순서를 유지하여 사용자가 처음 선택한 증상의 영양소가 먼저 노출되도록 한다.
    """
    seen: set[str] = set()
    ordered: list[str] = []
    for symptom in symptoms:
        for nutrient in SYMPTOM_TO_NUTRIENT.get(symptom, []):
            if nutrient not in seen:
                seen.add(nutrient)
                ordered.append(nutrient)
    return ordered


def collect_products(nutrients: Iterable[str]) -> list[Product]:
    """영양소 리스트에 매핑된 제품을 평탄화하여 반환한다."""
    seen: set[str] = set()
    result: list[Product] = []
    for nutrient in nutrients:
        for product in PRODUCTS.get(nutrient, []):
            key = product["name"]
            if key in seen:
                continue
            seen.add(key)
            result.append({**product, "nutrient": nutrient})  # type: ignore[typeddict-item]
    return result


def describe_nutrients(nutrients: Iterable[str]) -> list[dict]:
    return [
        {"name": n, "description": NUTRIENT_DESCRIPTION.get(n, "")} for n in nutrients
    ]


def recommend(symptoms: Iterable[str]) -> dict:
    nutrients = collect_nutrients(symptoms)
    products = collect_products(nutrients)
    return {
        "nutrients": describe_nutrients(nutrients),
        "products": products,
    }
